flips/context-propagation.md

```diff
@@ -0,0 +1,208 @@
+# Context Propagation and Network Layer Refactor (Core Protocol)
+
+
+
+## Objective
+
+FLIP to separate the API through which components are started from the API through which
+they expose their current status. 
+
+Also refactor the networking layer and to make it agnostic to Flow IDs.
+
+## Current Implementation
+
+The [`ReadyDoneAware`](https://github.com/onflow/flow-go/blob/7763000ba5724bb03f522380e513b784b4597d46/module/common.go#L6) interface provides an interface through which components / modules can be started and stopped. Calling the `Ready` method should start the component and return a channel that will close when startup has completed, and `Done` should be the corresponding method to shut down the component.
+
+When the network layer receives a message, it will pass the message to the [`Engine`](https://github.com/onflow/flow-go/blob/7763000ba5724bb03f522380e513b784b4597d46/network/engine.go) registered on
+the corresponding channel by [calling the engine's `Process` method](https://github.com/onflow/flow-go/blob/d31fd63eb651ed9faf0f677e9934baef6c4d9792/network/p2p/network.go#L406), passing it the Flow ID of the message sender.
+
+### Potential problems 
+#### Non-idempotency of `ReadyDoneAware`
+The current `ReadyDoneAware` interface is misleading, as by the name one might expect that it is only used to check the state of a component. However, in almost all current implementations the `Ready` method is used to both start the component *and* check when it has started up, and similarly for the `Done` method. 
```

Comment:
I think this poses a few difficulties.

First, if multiple components want to wait for the startup of another to complete (ie common dependency), then with this approach we would need to pass the channel returned from `Ready` to all of those components. This also means that we cannot construct those components until we've started the dependency, because we cannot get the ready channel until we call `Ready` to start it. Ideally, we don't want this restriction because it forces sequentialism of certain steps that could potentially run in parallel.

Now, what if these components also want to wait for the dependency's shutdown to complete later on? Then we have no way of doing this, because we only get a done channel when we call `Done` to shutdown the dependency. But these components have already been created by then, meaning they certainly don't have this done channel, and any mechanism we come up with of injecting it after wards is probably gonna be ugly.

Reply:
I thought about another solution, that might be easier to implement.

Since the goal is to make `ReadyDoneAware` idempotent, we could achieve by making `unit.Ready` and `unit.Done` idempotent, and each engine could wraps its Ready work with `unit.Ready`.

How to make `unit.Ready` idempotent?
We could cache the ready channel. If Ready is called more than once, then the cached ready channel will be returned. This way the second call will return the same ready channel and won't trigger additional initialization work.

In fact, it doesn't make sense to me that a module would initialize more than once, and I don't think there is such case exist now. 

So what I'm proposing is to just adjust `unit.Ready`:
```golang
func (u *Unit) Ready(checks ...func()) <-chan struct{} {
        // concurrent call will be blocking by the lock, this ensures only the first process will 
        // call the `checks` function, the second process will wait, and get notified once the first 
        // process finish all the checks.
        u.Lock()
        defer u.Unlock()

        if u.ready != nil {
                return u.ready
        }
            
	ready := make(chan struct{})
	go func() {
		for _, check := range checks {
			check()
		}
		close(ready)
	}()

        c.ready = ready
	return ready
}

```

In your example, A could be constructed with B whose Ready has been called by its owner, and A can call B's Ready again, which will not trigger another initialization, but wait on the previous initialization call to finish

Reply:
I think this is one valid solution, and I agree it may be easier to implement given the current state of the codebase. However, I would argue that if I were to design the API from scratch today, I would design it the way described in this FLIP, for all the reasons described in the "Motivations" section. Using contexts to start / stop components is so much more idiomatic in Go, and makes the APIs much more self explanatory. It also makes the separation of concerns more explicit.